src/Common/HashTable/HashTableKeyHolder.h

```diff
@@ -88,8 +88,12 @@ inline StringRef & ALWAYS_INLINE keyHolderGetKey(DB::ArenaKeyHolder & holder)
 
 inline void ALWAYS_INLINE keyHolderPersistKey(DB::ArenaKeyHolder & holder)
 {
-    // Hash table shouldn't ask us to persist a zero key
-    assert(holder.key.size > 0);
+    // Normally, our hash table shouldn't ask us to persist a zero key,
+    // but it can happened in the case of clearable hash table (ClearableHashSet, for example).
+    // The clearable hash table doesn't use zero storage and
+    // distinguishes empty keys by using cell version, not the value itself.
+    // So, when an empty StringRef is inserted in ClearableHashSet we'll get here key of zero size.
+    // assert(holder.key.size > 0);
     holder.key.data = holder.pool.insert(holder.key.data, holder.key.size);
```

Comment:
Technically, allocating zero sized data in the arena shouldn't hurt, the only a bit confusing part is that we can have different data pointing to the same address in such a case (AFAIU)